src/client.rs

```diff
@@ -1912,31 +1954,107 @@ impl TryFrom<&ChallengeRef<'_>> for BearerChallenge {
-/// fallback to manually hashing this content. This should only be done if the
-/// response body contains the image manifest.
-fn digest_header_value(headers: HeaderMap, body: Option<&[u8]>) -> Result<String> {
-    let digest_header = headers.get("Docker-Content-Digest");
-    match digest_header {
-        None => {
-            if let Some(body) = body {
-                // Fallback to hashing payload (tested with ECR)
-                let digest = sha2::Sha256::digest(body);
-                let hex = format!("sha256:{:x}", digest);
-                debug!(%hex, "Computed digest of manifest payload.");
-                Ok(hex)
-            } else {
-                Err(OciDistributionError::RegistryNoDigestError)
+/// Helper for extracting `Docker-Content-Digest` header from manifest GET or HEAD request.
+fn digest_header_value(headers: HeaderMap) -> Result<Option<String>> {
+    headers
+        .get(DOCKER_DIGEST_HEADER)
+        .map(|hv| hv.to_str().map(|s| s.to_string()))
+        .transpose()
+        .map_err(|e| {
+            OciDistributionError::GenericError(Some(format!(
+                "Cannot convert Docker-Content-Digest header value to string: {}",
+                e
+            )))
+        })
+}
+
+/// Given the optional digest header value and digest of the reference, returns the digest of the
+/// content, validating that the digest of the content matches the proper digest
+fn validate_digest(
+    body: &[u8],
+    digest_header: Option<String>,
+    reference_digest: Option<&str>,
+) -> Result<String> {
+    match compare_and_validate_digests(digest_header, reference_digest)? {
+        Some(digest) => calculate_and_validate(body, &digest),
+        // If we have neither, just digest the body
+        None => Ok(sha256_digest(body)),
+    }
+}
+
+/// This function compares the possible digest header value with a digest of a reference and returns
+/// the proper digest if it is valid. If both digests are provided, but they use different
+/// algorithms, then the header digest is returned as according to the spec it should be used for
+/// validation
+fn compare_and_validate_digests(
+    digest_header: Option<String>,
+    reference_digest: Option<&str>,
+) -> Result<Option<String>> {
+    match (digest_header, reference_digest) {
+        (None, None) => Ok(None),
+        (Some(digest), None) => Ok(Some(digest)),
+        (None, Some(digest)) => Ok(Some(digest.to_string())),
+        (Some(digest), Some(ref_digest)) => {
+            // According to the spec, it is just fine if the requested digest differs from the
+            // canonical digest from the headers if the algorithms are different. If they are the
+            // same, then the digests should match.
+            let (algo, _) = digest.split_once(':').ok_or_else(|| {
+                OciDistributionError::GenericError(Some(
+                    "Incorrect digest format on header value. Should have algorithm".to_string(),
+                ))
```

Comment:
Sorry for being a pain, but I would like to have dedicated error types. Otherwise it feels like going back to the early days when everything was just a `anyhow::Error`

Reply:
I did think about adding these errors, but the error enum is quite large. What if I added a new `digest.rs` file and had a separate enum for digest errors that would be included in the top level error type?

Reply:
That would be nice!